engine/runtime/src/main/java/org/enso/interpreter/runtime/error/WithWarnings.java

```diff
@@ -106,6 +106,20 @@ public WithWarnings append(EnsoContext ctx, ArrayRope<Warning> newWarnings) {
         value, ctx.getWarningsLimit(), warnings, newWarnings.toArray(Warning[]::new));
   }
 
+  // Ignore the warnings cache in .value and re-fetch them using the WarningsLibrary.
+  // This is only used for shouldWrap=true.
+  public Warning[] getWarningsNoCache(WarningsLibrary warningsLibrary) {
+    if (warningsLibrary != null && warningsLibrary.hasWarnings(value)) {
+      try {
+        return warningsLibrary.getWarnings(value, null, true);
+      } catch (UnsupportedMessageException e) {
+        throw EnsoContext.get(warningsLibrary).raiseAssertionPanic(warningsLibrary, null, e);
+      }
+    } else {
+      return Warning.fromSetToArray(warnings);
+    }
+  }
```

Comment:
I'm not sure if I understand.

I assume the problem scenario of this bug was:
1. warnings are fetched without wrapping
2. the result is cached
3. warnings are fetched with wrapping, but the cached unwrapped value is wrongly returned

Is my understanding correct?

----

I'm not sure if I get why the special handling _only_ applies for `shouldWrap=true`.

What if I first fetch warnings with wrapping and _that_ gets cached? Won't a subsequent fetch of warnings without wrapping also mistakenly get wrapped warnings because of them getting cached? Shouldn't we somehow keep separate caches for either case?

Reply:
I don't think that the caching happens during the call to `Warning.get_all`. `shouldWrap=True` only happens during `get_all`, so I don't think the possibility you describe is possible. `shouldWrap` is always false during normal evaluation and propagation. The `WithWarnings` caching happens during evaluation, and whether or not it happens seems very dependent on specialization details -- or, in this case, setting the environment variables that cause this bug to be reproducible. `WithWarnings` isn't just straightforward memoization of `getWarnings()`, it's just something that is constructed in some evaluation paths and not others.

The warnings stored in `WithWarnings` are gathered by the normal warning propagation that happens during evaluation, and wrapping never happens during that propagation, because wrapping is specific to the relationship between a particular container and its contents, and normal propagation just copies / transfers warnings to every downstream value. The wrapping only happens after evaluation and after warnings are attached to values, while traversing the resulting value for `get_all`, and so we have special-case handling for array-likes during this traversal, to add the wrappers.

At some point we may need to redesign the Warning propagation so that warnings describe not only what went wrong (the warning itself), but also information about the context it which it happened (the wrapper). The solution we have for this right now works, but if in the future we want to get more useful information to the user, we might need to add more logic to the propagation itself, such as where the warning came from. Then this special-case path won't be needed.

Reply:
I see, thanks for the explanation.

This seems to be very complicated indeed, I'm concerned that could bite us - it may be worth to try to take a step back and see if the logic could be overall simplified (when we have some time...). Maybe the way we want the warnings to behave just _is_ complex so maybe there isn't that much that we can do to simplify it, I don't know.

For now your explanation seems to imply that the fix is OK, so I'm ok with that.